src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs

```diff
@@ -19,88 +17,112 @@ namespace Microsoft.CodeAnalysis;
+                // Otherwise, merge the changes and set all the linked files to that merged content.
+                var mergeGroupResult = await MergeLinkedDocumentGroupAsync(newDocumentsAndHashes, sessionInfo, mergeConflictHandler, cancellationToken).ConfigureAwait(false);
+                linkedFileMergeResults.Add(mergeGroupResult);
+                updatedSolution = updatedSolution.WithDocumentTexts(
+                    relatedDocuments.SelectAsArray(d => (d, mergeGroupResult.MergedSourceText)));
             }
-
-            updatedSolution = updatedSolution.WithDocumentTexts(
-                allLinkedDocuments.SelectAsArray(documentId => (documentId, mergedText)));
         }
 
         return new LinkedFileMergeSessionResult(updatedSolution, linkedFileMergeResults);
     }
 
     private async Task<LinkedFileMergeResult> MergeLinkedDocumentGroupAsync(
-        ImmutableArray<DocumentId> allLinkedDocuments,
-        IGrouping<string, DocumentId> linkedDocumentGroup,
+        List<(Document newDocument, ImmutableArray<byte> newContentHash)> newDocumentsAndHashes,
         LinkedFileDiffMergingSessionInfo sessionInfo,
-        IMergeConflictHandler mergeConflictHandler,
+        IMergeConflictHandler? mergeConflictHandler,
         CancellationToken cancellationToken)
     {
+        Contract.ThrowIfTrue(newDocumentsAndHashes.Count < 2);
+
         var groupSessionInfo = new LinkedFileGroupSessionInfo();
 
         // Automatically merge non-conflicting diffs while collecting the conflicting diffs
 
         var textDifferencingService = oldSolution.Services.GetRequiredService<IDocumentTextDifferencingService>();
+
+        var firstNewDocument = newDocumentsAndHashes[0].newDocument;
+        var firstOldDocument = oldSolution.GetRequiredDocument(firstNewDocument.Id);
+        var firstOldSourceText = await firstOldDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+
         var appliedChanges = await textDifferencingService.GetTextChangesAsync(
-            oldSolution.GetDocument(linkedDocumentGroup.First()), newSolution.GetDocument(linkedDocumentGroup.First()), TextDifferenceTypes.Line, cancellationToken).ConfigureAwait(false);
-        var unmergedChanges = new List<UnmergedDocumentChanges>();
+            firstOldDocument, firstNewDocument, TextDifferenceTypes.Line, cancellationToken).ConfigureAwait(false);
 
-        foreach (var documentId in linkedDocumentGroup.Skip(1))
+        var unmergedChanges = new List<UnmergedDocumentChanges>();
+        for (int i = 1, n = newDocumentsAndHashes.Count; i < n; i++)
         {
+            var siblingNewDocument = newDocumentsAndHashes[i].newDocument;
+            var siblingOldDocument = oldSolution.GetRequiredDocument(siblingNewDocument.Id);
+
             appliedChanges = await AddDocumentMergeChangesAsync(
-                oldSolution.GetDocument(documentId),
-                newSolution.GetDocument(documentId),
-                [.. appliedChanges],
```

Comment:
this was converting an IA into a list, that was never mutated.  made the input/output entirely IA bases.